### ISSUE TITLE
Epic: Testing

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3046]
As we crate smash and push towards 1.0 should we make explicit our testing goals?

I personally would like to put more time into testing, but do not know the priority, my vague idea is as we get a crate near to 1.0 it needs more thought put into testing it. 

On kix's advice I read dpc's post about testing, using his term I think we want to calcify the 1.0 crates.

ref: https://github.com/rust-bitcoin/rust-bitcoin/discussions/2977

**Comment by tcharding** [comment]
cc @dpc and @Kixunil

**Comment by tcharding** [comment]
This also relates to mutation testing, which in my brief experience led me to write 100s of unit tests that test the internals of a function. Perhaps I'm doing it wrong?

**Comment by dpc** [comment]
> ref: https://dpc.pw/posts/be-warned-tests-are-great-calcifiers/

Hey, that's a good piece I forgot I wrote. :D


1.0 means calcification, so there's no need to have reservations, and it's time to go all in on testing public APIs. In particular in Bitcoin - the logic under tests is not going to change easily.

rust-bitcoin, being a foundational utility library is somewhat wide and shallow, AFAICT. It has a lot of pieces of functionality that do not necessarily overlap or even interact, and in itself is often a bottom layer of larger applications.

So it should be relatively easy to take each piece of functionality and wrap it in some unit-tests that give confidence that it is doing the right thing. In an application code, like Fedimint, there are layers and layers of logic, usually with side effects - testing that is much greater challange, and requires more e2e tests.

One challenge is that the code like this often has very broad state, and things tend to break in important edge cases. So property testing and fuzz testing are particularly valuable, as they can detect edge cases automatically.


> This also relates to mutation testing, which in my brief experience led me to write 100s of unit tests that test the internals of a function. Perhaps I'm doing it wrong?

I really like the concept of mutation testing, but I haven't ever really got to implement it in Rust project. I'm not aware of any tools etc. that would make it easier. I'd be happy to learn on sombody elses experiences here. :)

Mutation testing starts by discovering which bugs would go unnoticed. But then how exactly to cover such cases is the issue and the skill. Generating tons and tons of tests cases that cover every turn and twist in the code will lead to maintenance burden. If instead some general higher level properties or at least a table of "test cases" can be distilled, property testing and fuzz testing can cover everything with minimal 

Unfortunately code in Bitcoin tends to have quite hairy internal properties and a lot of edge cases, so the properties and assertion there might also be quite complex. But still - it's important to try to cover as much as possible, with few generalized tests, as much as possible.

@tcharding

**Comment by shinghim** [comment]
I started reading up on testing in Rust from looking at #2401. Seems like they're pretty closely related - should we close that issue and roll it into this one?

**Comment by tcharding** [comment]
Did you link to the issue you meant to @shinghim?

**Comment by tcharding** [comment]
Thanks @dpc, you the man. I'll definitely cc you if/when I get to mutation testing.

**Comment by shinghim** [comment]
I thought it might be related since there's some discussion of property testing in that issues as well, albeit for a more specific case. I just wanted to see if there was any overlap in scope in case it causes discussion of the same thing in two places, but either way, I'd like to help out on the testing so I'll keep reading up on it!

**Comment by tcharding** [comment]
No worries, I don't think this is a duplicate but yes it could be seen as related. Thanks, help most appreciated!

**Comment by Kixunil** [comment]
> I think we want to calcify the 1.0 crates.

Yes, the API and whatever promises we make in the docs, not the internals.

> This also relates to mutation testing, which in my brief experience led me to write 100s of unit tests that test the internals of a function. Perhaps I'm doing it wrong?

I believe you should be able to test those things purely from public API. If you cannot, something weird is going on and I want to see such case to understand it.

> 1.0 means calcification, so there's no need to have reservations, and it's time to go all in on testing public APIs. In particular in Bitcoin - the logic under tests is not going to change easily.

Yes, that's exactly what I was thinking.

> Mutation testing starts by discovering which bugs would go unnoticed. But then how exactly to cover such cases is the issue and the skill. Generating tons and tons of tests cases that cover every turn and twist in the code will lead to maintenance burden. If instead some general higher level properties or at least a table of "test cases" can be distilled, property testing and fuzz testing can cover everything with minimal burden.

I believe this is accurate.
